sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala

```
@@ -0,0 +1,142 @@
+        // the schema validation doesn't break even it fails, though it might miss on detecting
+        // change which is not a big deal.
+        createSchemaFile(keySchema, valueSchema)
+      }
+    } else {
+      // schema doesn't exist, create one now
+      logDebug(s"Schema file for provider $providerId doesn't exist. Creating one.")
+      createSchemaFile(keySchema, valueSchema)
+    }
+  }
+
+  private def schemasCompatible(storedSchema: StructType, schema: StructType): Boolean =
+    equalsIgnoreCompatibleNullability(storedSchema, schema)
+
+  private def equalsIgnoreCompatibleNullability(from: DataType, to: DataType): Boolean = {
+    // This implementations should be same with DataType.equalsIgnoreCompatibleNullability, except
+    // this shouldn't check the name equality.
+    (from, to) match {
+      case (ArrayType(fromElement, fn), ArrayType(toElement, tn)) =>
+        (tn || !fn) && equalsIgnoreCompatibleNullability(fromElement, toElement)
+
+      case (MapType(fromKey, fromValue, fn), MapType(toKey, toValue, tn)) =>
+        (tn || !fn) &&
+          equalsIgnoreCompatibleNullability(fromKey, toKey) &&
+          equalsIgnoreCompatibleNullability(fromValue, toValue)
+
+      case (StructType(fromFields), StructType(toFields)) =>
+        fromFields.length == toFields.length &&
+          fromFields.zip(toFields).forall { case (fromField, toField) =>
+              (toField.nullable || !fromField.nullable) &&
+              equalsIgnoreCompatibleNullability(fromField.dataType, toField.dataType)
+          }
+
+      case (fromDataType, toDataType) => fromDataType == toDataType
+    }
+  }
+
+  private def readSchemaFile(): (StructType, StructType) = {
+    val inStream = fm.open(schemaFileLocation)
+    try {
+      val version = inStream.readInt()
+      // Currently we only support version 1, which we can simplify the version validation and
+      // the parse logic.
+      require(version == StateSchemaCompatibilityChecker.VERSION,
```

This is not an illegal argument. It's better to throw UnsupportedException.

---

Let me leverage HDFSMetadataLog instead.